server/src/l10n.js

```diff
@@ -0,0 +1,85 @@
+const fs = require("fs");
+const path = require("path");
+const globby = require("globby");
+require("fluent-intl-polyfill");
+const { MessageContext } = require("fluent");
+const negotiateLanguages = require("fluent-langneg/compat");
+const mozlog = require("./logging").mozlog("l10n");
+
+let userLangs;
+const contexts = [];
+let inited = false;
+
+function init(userLocales = []) {
+  if (inited) {
+    return Promise.resolve();
+  }
+  inited = true;
+  userLangs = userLocales;
+  return _getLocales(userLocales).then(locales => {
+    // module-global assignment
+    userLangs = locales;
+
+    const mcPromises = [];
+    userLangs.forEach(lang => {
+      const mc = new MessageContext(lang);
+      mcPromises.push(new Promise((resolve, reject) => {
+        // TODO: is this path correct?
+        const filename = path.join(__dirname, '..', '..', 'locales', lang, 'server.ftl');
+        fs.readFile(filename, 'utf-8', (err, data) => {
+          if (err) { return reject(err); }
+          mc.addMessages(data);
+          contexts[lang] = mc;
+          resolve(mc);
+        });
+      }));
+    });
+    return Promise.all(mcPromises);
+  }).catch(err => {
+    mozlog.error('l10n-init-failed', {err: err});
+  });
+}
+
+function getText(l10nID, args) {
+  if (!inited) { init(); }
+  // Find the first MessageContext with the l10n ID, in order of user preference.
+  for (let lang of userLangs) {
+    if (contexts[lang].hasMessage(l10nID)) {
+      return contexts[lang].getMessage(l10nID, args);
+    }
+  }
+  return null;
+}
+
+function _getLocales(requestedLocales) {
+  return _getAvailableLocales().then(availableLocales => {
+    const locales = negotiateLanguages(
+      userLangs,
+      availableLocales,
+      { defaultLocale: 'en-US' }
+    );
+    return locales;
+  });
+}
+
+module.exports = { init, getText, userLangs, generateMessages };
```

Comment:
Generally the code uses a pattern of setting `exports.init` etc directly – there's a downside of potentially having to refer to `exports.init()` locally in the module, but otherwise I like seeing that something is public right where it is declared.